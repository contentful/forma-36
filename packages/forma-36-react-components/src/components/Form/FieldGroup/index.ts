export * from './FieldGroup';
export { default } from './FieldGroup';
