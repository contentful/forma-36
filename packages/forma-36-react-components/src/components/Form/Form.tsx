import React, { FormEventHandler, FormEvent } from 'react';
import cn from 'classnames';

import styles from './Form.css';

export interface FormProps {
  onSubmit?: FormEventHandler;
  spacing?: 'condensed' | 'default';
  testId?: string;
  style?: React.CSSProperties;
  className?: string;
  children: React.ReactChild | React.ReactNodeArray;
}

const defaultProps: Partial<FormProps> = {
  spacing: 'default',
  testId: 'cf-ui-form',
};

export const Form = (props: FormProps) => {
  const {
    className,
    children,
    testId,
    onSubmit,
    spacing,
    ...otherProps
  } = props;

  const classNames = cn(styles.Form, className);

  const formItemClassNames = cn(
    styles.Form__item,
    styles[`Form__item--${spacing}`],
  );

  const handleSubmit = (event: FormEvent) => {
    event.preventDefault();
    if (onSubmit) {
      onSubmit(event);
    }
  };

  return (
    <form
      className={classNames}
      data-test-id={testId}
      onSubmit={handleSubmit}
      {...otherProps}
    >
      {React.Children.map(children, (child) => {
        if (child) {
          return <div className={formItemClassNames}>{child}</div>;
        }
        return null;
      })}
    </form>
  );
};

Form.defaultProps = defaultProps;

export default Form;
