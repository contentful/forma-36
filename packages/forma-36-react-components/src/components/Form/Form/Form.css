@import 'settings/typography';
@import 'settings/dimensions';

.Form {
  display: block;
}

.Form__item {
  display: block;
  margin-bottom: var(--spacing-l);
}
