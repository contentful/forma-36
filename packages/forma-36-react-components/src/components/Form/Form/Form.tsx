import React, { Component, FormEventHandler, FormEvent } from 'react';
import cn from 'classnames';

const styles = require('./Form.css');

export type FormProps = {
  onSubmit?: FormEventHandler;
  testId?: string;
  style?: React.CSSProperties;
  className?: string;
  children: React.ReactChild | React.ReactNodeArray;
} & typeof defaultProps;

const defaultProps = {
  testId: 'cf-ui-form',
};

export class Form extends Component<FormProps> {
  static defaultProps = defaultProps;

  handleSubmit = (event: FormEvent) => {
    event.preventDefault();
    if (this.props.onSubmit) {
      this.props.onSubmit(event);
    }
  };

  render() {
    const { className, children, testId, onSubmit, ...otherProps } = this.props;

    const classNames = cn(styles.Form, className);

    return (
      <form
        className={classNames}
        data-test-id={testId}
        onSubmit={this.handleSubmit}
        {...otherProps}
      >
        {React.Children.map(children, child => {
          if (child) {
            return <div className={styles.Form__item}>{child}</div>;
          }
          return null;
        })}
      </form>
    );
  }
}

export default Form;
