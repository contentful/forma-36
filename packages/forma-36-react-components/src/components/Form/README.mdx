---
title: 'Form'
type: 'component'
status: 'stable'
slug: /components/form/
github: 'https://github.com/contentful/forma-36/tree/master/packages/forma-36-react-components/src/components/Form'
storybook: 'https://f36-storybook.contentful.com/?path=/story/components-form--default'
---

`Form` is a wrapper component to apply a consistent layout and handle form submissions.

## Table of contents

- [How to use Form](#how-to-use-Form)
- [Component variations](#component-variations)
- [Content recommendations](#content-recommendations)
- [Accessibility](#accessibility)

## How to use Form

- The `Form` component should wrap FieldGroup components to apply consistent layouts
- Provide an `onSubmit` handler for form validation and submission

## Component variations

### Basic example

```jsx
<React.Fragment>
  <Form onSubmit={() => console.log('onSubmit')} spacing="default">
    <FieldGroup>
      <TextField
        required
        name="nameInput"
        id="nameInput"
        labelText="Name"
        value="Claus Mitchell"
        helpText="Provide a full name"
      />
      <TextField
        required
        name="emailInput"
        id="emailInput"
        labelText="Email address"
        value="claus.mitchell@contentful.com"
        helpText="Provide an email address"
      />
    </FieldGroup>
    <FieldGroup>
      <Button>Submit</Button>
    </FieldGroup>
  </Form>
</React.Fragment>
```

### Default spacing (`spacing="default"`)

```jsx
<React.Fragment>
  <Form onSubmit={() => console.log('onSubmit')} spacing="default">
    <FieldGroup>
      <TextField
        required
        name="nameInput"
        id="nameInput"
        labelText="Name"
        value="Claus Mitchell"
        helpText="Provide a full name"
      />
      <TextField
        required
        name="emailInput"
        id="emailInput"
        labelText="Email address"
        value="claus.mitchell@contentful.com"
        helpText="Enter email address"
      />
    </FieldGroup>
    <FieldGroup>
      <Button>Submit</Button>
    </FieldGroup>
  </Form>
</React.Fragment>
```

### Condensed spacing (`spacing="condensed"`)

```jsx
<React.Fragment>
  <Form onSubmit={() => console.log('onSubmit')} spacing="condensed">
    <FieldGroup>
      <TextField
        required
        name="nameInput"
        id="nameInput"
        labelText="Name"
        value="Claus Mitchell"
        helpText="Provide a full name"
      />
      <TextField
        required
        name="emailInput"
        id="emailInput"
        labelText="Email address"
        value="claus.mitchell@contentful.com"
        helpText="Enter email address"
      />
    </FieldGroup>
    <FieldGroup>
      <Button>Submit</Button>
    </FieldGroup>
  </Form>
</React.Fragment>
```

## Content recommendations

Missing

## Accessibility

Missing
