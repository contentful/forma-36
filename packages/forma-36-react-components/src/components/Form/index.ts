export { Form } from './Form';
export type { FormProps } from './Form';
export { FieldGroup } from './FieldGroup/FieldGroup';
export type { FieldGroupProps } from './FieldGroup/FieldGroup';
