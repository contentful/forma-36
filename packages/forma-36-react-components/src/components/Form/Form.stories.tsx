import React, { useState } from 'react';

import Form, { FormProps } from './Form';
import FieldGroup from './FieldGroup';
import Button from '../Button';
import TextField from '../TextField';
import CheckboxField from '../CheckboxField';
import notes from './README.mdx';

export default {
  title: 'Form Elements/Form',
  component: Form,
  parameters: {
    propTypes: [Form['__docgenInfo']],
    notes,
  },
  argTypes: {
    className: { control: { disable: true } },
    testId: { control: { disable: true } },
  },
};

export const Basic = (args: FormProps) => {
  const [agreeTerms, setTerms] = useState('yes');
  return (
    <Form {...args}>
      <TextField
        required
        name="nameInput"
        id="nameInput"
        labelText="Name"
        value="Johannes Bugiel"
        helpText="Please enter your name"
      />
      <TextField
        required
        name="emailInput"
        id="emailInput"
        labelText="E-Mail"
        value="johannes.bugiel@contentful.com"
        helpText="Please enter your mail"
      />
      <FieldGroup>
        <CheckboxField
          labelText="I agree"
          value="yes"
          helpText="Click if you agree"
          onChange={(e) => setTerms((e.target as HTMLInputElement).value)}
          checked={agreeTerms === 'yes'}
          id="termsCheckboxYes"
        />
        <CheckboxField
          labelText="I don't agree"
          value="no"
          onChange={(e) => setTerms((e.target as HTMLInputElement).value)}
          checked={agreeTerms === 'no'}
          helpText="Click if you don't agree"
          id="termsCheckboxNo"
        />
      </FieldGroup>
      <Button>Submit</Button>
    </Form>
  );
};

Basic.args = {
  spacing: 'default',
};
