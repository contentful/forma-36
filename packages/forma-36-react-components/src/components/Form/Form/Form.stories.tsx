import React from 'react';
import { storiesOf, StoryDecorator } from '@storybook/react';
import { StateDecorator, Store } from '@sambego/storybook-state';
import { action } from '@storybook/addon-actions';
import { boolean, select } from '@storybook/addon-knobs';

import Form from './Form';
import FieldGroup from '../FieldGroup';
import Button from '../../Button';
import TextField from '../../TextField';
import CheckboxField from '../../CheckboxField';

const store = new Store({
  agreeTerms: 'yes',
});

storiesOf('Components|Form', module)
  .addDecorator(StateDecorator(store) as StoryDecorator)
  .add('default', () => (
    <Form
      onSubmit={action('onSubmit')}
      spacing={select(
        'Spacing',
        {
          Default: 'default',
          Condensed: 'condensed',
        },
        'default',
      )}
    >
      <TextField
        required
        name="nameInput"
        id="nameInput"
        labelText="Name"
        value="Johannes Bugiel"
        helpText="Please enter your name"
      />
      <TextField
        required
        name="emailInput"
        id="emailInput"
        labelText="E-Mail"
        value="johannes.bugiel@contentful.com"
        helpText="Please enter your mail"
      />
      <FieldGroup row={boolean('Field Group Row', false)}>
        <CheckboxField
          labelText="I agree"
          value="yes"
          helpText="Click if you agree"
          onChange={e =>
            store.set({ agreeTerms: (e.target as HTMLInputElement).value })
          }
          checked={store.state.agreeTerms === 'yes'}
          id="termsCheckboxYes"
        />
        <CheckboxField
          labelText="I don't agree"
          value="no"
          onChange={e =>
            store.set({ agreeTerms: (e.target as HTMLInputElement).value })
          }
          checked={store.state.agreeTerms === 'no'}
          helpText="Click if you don't agree"
          id="termsCheckboxNo"
        />
      </FieldGroup>
      <Button>Submit</Button>
    </Form>
  ));
