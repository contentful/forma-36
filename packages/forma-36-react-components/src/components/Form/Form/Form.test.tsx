import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import Form from './Form';
import TextField from '../../TextField';

it('renders the component', () => {
  const output = shallow(
    <Form>
      <TextField
        required
        name="nameInput"
        id="nameInput"
        labelText="Name"
        value="Johannes Bugiel"
        helpText="Please enter your name"
      />
    </Form>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <Form className="my-extra-class">
      <TextField
        required
        name="nameInput"
        id="nameInput"
        labelText="Name"
        value="Johannes Bugiel"
        helpText="Please enter your name"
      />
    </Form>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with condensed spacing', () => {
  const output = shallow(
    <Form className="my-extra-class">
      <TextField
        required
        name="nameInput"
        id="nameInput"
        labelText="Name"
        value="Johannes Bugiel"
        helpText="Please enter your name"
      />
    </Form>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with condensed spacing (prevent default)', () => {
  const mockOnSubmit = jest.fn();
  const output = shallow(
    <Form className="my-extra-class" onSubmit={mockOnSubmit}>
      <TextField
        required
        name="nameInput"
        id="nameInput"
        labelText="Name"
        value="Johannes Bugiel"
        helpText="Please enter your name"
      />
    </Form>,
  );

  output.simulate('submit', {
    preventDefault: () => {},
  });
  expect(mockOnSubmit).toHaveBeenCalled();
});

it('has no a11y issues', async () => {
  const output = mount(
    <Form>
      <TextField
        required
        name="nameInput"
        id="nameInput"
        labelText="Name"
        value="Johannes Bugiel"
        helpText="Please enter your name"
      />
    </Form>,
  ).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
