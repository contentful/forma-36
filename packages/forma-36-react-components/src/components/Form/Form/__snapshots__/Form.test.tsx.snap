// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<form
  className="Form"
  data-test-id="cf-ui-form"
  onSubmit={[Function]}
>
  <div
    className="Form__item"
    key=".0"
  >
    <TextField
      countCharacters={false}
      helpText="Please enter your name"
      id="nameInput"
      labelText="Name"
      name="nameInput"
      required={true}
      testId="cf-ui-text-field"
      textarea={false}
      value="Johannes Bugiel"
    />
  </div>
</form>
`;

exports[`renders the component with an additional class name 1`] = `
<form
  className="Form my-extra-class"
  data-test-id="cf-ui-form"
  onSubmit={[Function]}
>
  <div
    className="Form__item"
    key=".0"
  >
    <TextField
      countCharacters={false}
      helpText="Please enter your name"
      id="nameInput"
      labelText="Name"
      name="nameInput"
      required={true}
      testId="cf-ui-text-field"
      textarea={false}
      value="Johannes Bugiel"
    />
  </div>
</form>
`;

exports[`renders the component with condensed spacing 1`] = `
<form
  className="Form my-extra-class"
  data-test-id="cf-ui-form"
  onSubmit={[Function]}
>
  <div
    className="Form__item"
    key=".0"
  >
    <TextField
      countCharacters={false}
      helpText="Please enter your name"
      id="nameInput"
      labelText="Name"
      name="nameInput"
      required={true}
      testId="cf-ui-text-field"
      textarea={false}
      value="Johannes Bugiel"
    />
  </div>
</form>
`;
